man/ffpmerge.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpmerge 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpmerge - This program merges the rows of an FFP into a single row.
.SH SYNOPSIS
.BI "ffpmerge [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
The behavior of the program is to merge all rows of the FFP.  The
FFP can be a columnar or key-valued FFP.
FFPs will be read from standard input if no options are supplied and 
.B ffpmerge 
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list.

.SH OPTIONS
.TP 
.B -d, --disable
Disable classing of amino acid and nucleotide sequences. 
.TP
.B -k, --keys
Print key value pairs. Default is output in columnar format.
.TP
.B -a, --amino
Input FFPs are animo acid sequences.
.TP
.B -t, --text
Input FFPs are text.
.TP
.B -v, --version
Display version.
.TP
.B -h, --help
Display help message.
.PP
.SH EXAMPLES
The main usage of this program is to merge separately calculated
FFPs from different parts of a large genome.  For example consider
a multi-chromosome genome.
.PP
.CODE ffpry -l 5 chr1.fna > chr1.vector
.CODE ffpry -l 5 chr2.fna > chr2.vector
.CODE ffpmerge chr*.vector > chrall.vector
.PP
The primary advantage of computing the vectors in this way
is that separate chromosome vectors can be calculated on
different processes and the process of 
.I "l" "-mer" 
counting can be distributed across a multi-processor
architecture.  If you have many small fasta files which
you need to merge there is no clear advantages to using the
count and merge process shown above. In this case, concatenating
the files together on the fly is more effective:
.PP
.CODE cat *.fna | ffpry -l 5 > merged.vector
.PP
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpaa(1),
.BR ffpry(1),
.BR ffprwn(1),
.BR ffpjsd(1),
.BR ffpcol(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpmerge.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>
#include "hash.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "../config.h"

char PROG_NAME[FILENAME_MAX];
#define COL_SIZE 1000 /**< Intial guess of column number in the FP */
#define MAX_WORD_LENGTH 40 /**< Maximum feature length */

unsigned merge(FILE * fp, unsigned **vals, unsigned size);
void mergeHash(FILE * fp);
void mergeHashComplement(FILE * fp);
int isKeyBased(FILE * fp);
int getKeyLength(FILE * fp);


char usage_str[] = "Usage: %s [OPTIONS] ... [FILE]\n\
This program merges multiple rows of an FFP into a single row\n\n\
The behavior of the program is to merge all rows of the FFP.\n\
The FFP can be a columnar or key-valued FFP.\n\
\t-t, --text\tInput is text\n\
\t-a, --amino\tInput is amino acid\n\
\t-d, --disable\tDisable classing of input\n\
\t-k, --keys\tPrint key-value pairs.\n\
\t-c, --complement\tMerges keys with rev complement (must have -d).\n\
\t-v, --vesion\tDisplay version\n\
\t-h, --help\tThis messsage\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


char *weightVector;
int Length = MAX_WORD_LENGTH;
			    /**< Feature Length */

int main(int argc, char **argv)
{
    FILE *fp;
    int opt;
    unsigned cols;
    unsigned i;
    unsigned *vals;
    char **keys;
    bool isKey = false;
    bool dflag = false;
    bool aflag = false;
    bool tflag = false;
    bool kflag = false;
    bool cflag = false;
    int option_index = 0;


    static struct option long_options[] = {
	{"help", no_argument, 0, 'h'},
	{"disable", no_argument, 0, 'd'},
	{"keys", no_argument, 0, 'k'},
	{"amino", no_argument, 0, 'a'},
	{"text", no_argument, 0, 't'},
	{"complement", no_argument, 0, 'c'},
	{"version", no_argument, 0, 'v'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();

    strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "khdactv",
			      long_options, &option_index)) != -1)
	switch (opt) {
	case 'd':
	    dflag = !dflag;
	    break;
	case 'a':
	    aflag = !aflag;
	    break;
	case 't':
	    tflag = !tflag;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'k':
	    kflag = !kflag;
	    break;
	case 'c':
	    cflag = !cflag;
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    break;
	}

    if (aflag && dflag)
	fatal_msg("Error -a or -t, not both\n");

    if (cflag && !dflag)
	fatal_msg("Error -c must be used with -d");


    if (aflag)
	initHash(0, amino, !dflag);
    else if (tflag)
	initHash(0, text, !dflag);
    else
	initHash(0, nucleotide, !dflag);

// init to zero
    vals = (unsigned *) calloc(sizeof(unsigned), COL_SIZE);


// IT may be necessary to determine what kind of data is 
// being merged, add additional switches -disable-classes?

// masking should also be used in this function

    cols = COL_SIZE;


    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
	    	fatal_msg("%s: %s\n", *argv, strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	isKey = isKeyBased(fp);

	if (!isKey)
	    cols = merge(fp, &vals, cols);
	else {
	    Length = getKeyLength(fp);
	    if (cflag)
	        mergeHashComplement(fp);
	    else
	        mergeHash(fp);
	}

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);

    if (!isKey) {
	for (i = 0; i < cols-1; i++)
	    printf("%u\t", vals[i]);
	printf("%u\n", vals[i]);
    } else {
	hashKeys(&keys);
	if (kflag) {
	    for (i = 0; i < numKeys()-1; i++)
		printf("%s %u\t", keys[i], hashval(keys[i]));
	    printf("%s %u\n", keys[i], hashval(keys[i]));
	} else {
	    for (i = 0; i < numKeys()-1; i++)
		printf("%u\t", hashval(keys[i]));
	    printf("%u\n", hashval(keys[i]));
	}
	//printf("\n");
	freeHash();
	//@todo Free keys?
    }

    free(vals);
    return EXIT_SUCCESS;
}

/**
 * Merges all the rows in a single column together
 *
 * All the frequencies in a column are added together
 * to the vector in vals.  The FFP in the file fp
 * must be stored in columnar format.  Space in vals
 * is dynamically reallocated as needed to accomodate
 * the size of the FFP.
 *
 * @param fp A file pointer to an FFP
 * @param vals A ptr to an array of unsigned integers
 * @param size An initial guess as to the size of vals.
 * @return The number of columns read.
 */


unsigned merge(FILE * fp, unsigned **vals, unsigned size)
{
    long unsigned i;
    int rows;
    unsigned val;
    unsigned cols = 0;
    char ch[2];
    int d;


    i = 0;
    rows = 0;
    while (!feof(fp)) {
	if (i < size) {
	    d = fscanf(fp, "%u%[\n\r]", &val, ch);
	    (*vals)[i++] += val;
	    if (d == 2) {
		cols = i;
		rows++;
		i = 0;
	    }
	} else {
	    size += COL_SIZE;
	    *vals = (unsigned *) realloc(*vals, sizeof(unsigned) * size);
	}
    }
    return cols;

}


/**
 * Merges all the rows in a (key,value) FFP
 *
 * All the frequencies in a (key,value) FFP row are added together
 * in a new hash.  The FFP in the file fp
 * must be a (Key,value) FFP.
 *
 * @param fp A file pointer to an FFP
 * @return None
 */


void mergeHash(FILE * fp)
{
    unsigned val;
    char ch[2];
    char s[Length];


    while (!feof(fp)) {
	fscanf(fp, "%s%[\t]%u%[\n\r]", s, ch, &val, ch);
	s[Length] = '\0';	/**<This line may be unnecessary**/
	hashAdd(s, val);	// increments hash by value;
    }
}

/**
 * Merges all the rows in a (key,value) FFP and orders by hash key
 *
 * All the frequencies in a (key,value) FFP row are added together
 * in a new hash.  The FFP in the file fp
 * must be a (Key,value) FFP.
 *
 * @param fp A file pointer to an FFP
 * @return None
 */


void mergeHashComplement(FILE * fp)
{
    unsigned val;
    char ch[2];
    char s[Length];
    char *r; //reverse complement

    while (!feof(fp)) {
	fscanf(fp, "%s%[\t]%u%[\n\r]", s, ch, &val, ch);
	s[Length] = '\0';	/**<This line may be unnecessary**/

	// for fixing bug where first character read is not a valid letter
	if (!isalpha(s[0]))
	    continue;

	r = strdup(s, Length);
	r[Length] = '\0';
	rev(r,Length);
	complement(r,Length);

	// merge rev complemented values, store the string that's lowest
	//printf("s=%s,r=%s\n",s,r);

	if (strcmp(s,r) <= 0) {
    		hashAdd(s, val);	// increments hash by value;
	} else {
		hashAdd(r, val);
	}

	free(r);
    }
}
